lib/core/src/Ouroboros/Network/Client/Wallet.hs

```diff
@@ -475,62 +479,103 @@ type LocalStateQueryResult state = Either AcquireFailure state
+    clientStIdle = do
+        cmd <- awaitNextCmd
+        (era, pt) <- atomically getTip
+        pure $ LSQ.SendMsgAcquire pt (clientStAcquiring cmd pt era)
 
     clientStAcquiring
-        :: forall state. Query block state
-        -> (LocalStateQueryResult state -> m ())
+        :: LocalStateQueryCmd block m
+        -> Point block
+        -> AnyCardanoEra
         -> LSQ.ClientStAcquiring block (Point block) (Query block) m Void
-    clientStAcquiring query respond = LSQ.ClientStAcquiring
-        { recvMsgAcquired = clientStAcquired query respond
-        , recvMsgFailure = \failure -> do
-                respond (Left failure)
-                clientStIdle
+    clientStAcquiring qry oldPt era = LSQ.ClientStAcquiring
+        { recvMsgAcquired = clientStAcquired qry era
+        , recvMsgFailure = \_failure -> do
+            (newEra, newPt) <- atomically $ do
+                t <- getTip
+                guard (((snd t) /= oldPt))
+                return t
+            pure $ LSQ.SendMsgAcquire newPt (clientStAcquiring qry newPt newEra)
         }
 
     clientStAcquired
-        :: forall state. Query block state
-        -> (LocalStateQueryResult state -> m ())
-        -> LSQ.ClientStAcquired block (Point block) (Query block) m Void
-    clientStAcquired query respond =
-        LSQ.SendMsgQuery query (clientStQuerying respond)
-
-    -- By re-acquiring rather releasing the state with 'MsgRelease' it
-    -- enables optimisations on the server side.
-    clientStAcquiredAgain
-        :: m (LSQ.ClientStAcquired block (Point block) (Query block) m Void)
-    clientStAcquiredAgain = awaitNextCmd <&> \case
-        CmdQueryLocalState pt query respond ->
-            LSQ.SendMsgReAcquire pt (clientStAcquiring query respond)
-
-    clientStQuerying
-        :: forall state. (LocalStateQueryResult state -> m ())
-        -> LSQ.ClientStQuerying block (Point block) (Query block) m Void state
-    clientStQuerying respond = LSQ.ClientStQuerying
-        { recvMsgResult = \result -> do
-            respond (Right result)
-            clientStAcquiredAgain
-        }
+        :: LocalStateQueryCmd block m
+        -> AnyCardanoEra
+        -> (LSQ.ClientStAcquired block (Point block) (Query block) m Void)
+    clientStAcquired (SomeLSQ cmd respond) era = go cmd $ \res -> do
+        LSQ.SendMsgRelease (respond res >> clientStIdle)
+        -- TODO: We might not want to release. Re-acquiring, or even staying
+        -- acquired should be more efficient.
+      where
+          go
+              :: forall a. LSQ block m a
+              -> (a -> (LSQ.ClientStAcquired block (Point block) (Query block) m Void))
+              -> (LSQ.ClientStAcquired block (Point block) (Query block) m Void)
+          go (LSQPure a) cont = cont a
+          go (LSQry qry) cont = LSQ.SendMsgQuery qry $ LSQ.ClientStQuerying $ \res -> do
+              pure $ cont res
+              -- TODO: It would be nice to trace the time it takes to run the
```

Comment:
This does not appear to be straight forward to do. There is no opportunity to run `IO` _before_ the query, but _after_ the point is acquired.

Reply:
I re-added top-level `bracketQuery` in the network layer that time the full `send` action (includes acquiring, and all the included queries).